### PR TITLE
feat: Add SEDA Keys

### LINKED ISSUE #162: ✨ Pubkeys module
## Motivation

The pubkeys module would serve as a repository for various public keys used in the protocol. Having a separate module for this purpose has the following advantages:

- Provide a central repository of public keys that other modules can have access to
- Facilitate management of keys on various curves (For example, we need to implement a single transaction endpoint for registering keys of any type)
- Support key rotation (to be implemented later)

## Implementation

The store for this module would have the following key-value scheme:
```
validator | curve -> public_key
```

There would be a single transaction endpoint for registering a key. A flag `--rotate` may be added to request overwriting an existing key.
```
$ seda-chaind tx pubkeys register <valoper> <curve> <pubkey> --from <delegator_key>
```

### PR DESCRIPTION
**Pull request by hacheigriega**
## Explanation of Changes

This PR adds the last component of the pubkey module, the SEDA keys. This implements a CLI command for generating a set of SEDA keys and publishing their public keys on chain. Each of the SEDA keys are generated from a random seed, and they are stored in a single key file `seda_keys.json` under the same directory where the consensus private key is saved to. This means that, if any of the keys is lost, the user must re-generate and rotate all of his or her SEDA keys.

To generate the SEDA keys and register their public keys on chain, run:
```
$ sedad tx pubkey add-seda-keys --from satoshi --keyring-backend test --fees 10000000000000000aseda --gas auto --gas-adjustment 1.5
```

To query the validator's list of SEDA public keys:

```
$ sedad query pubkey validator-keys sedavaloper1nsnv4g2ejq36asujaettw4ge73pzzpjf86208z
validator_pub_keys:
  indexed_pub_keys:
  - index: 0
    pub_key:
      '@type': /cosmos.crypto.secp256k1.PubKey
      key: A7G7jKH0nldfGvWcLf+KudfJha4iYr4Gq5PRAXvnqVZA
  validator_addr: sedavaloper1nsnv4g2ejq36asujaettw4ge73pzzpjf86208z1000000 10000000000aseda                                                  
```

When the SEDA Protocol requires a new key to be registered on chain, the `add-key` command would be updated to generate the new key at some unused index. 

## Related PRs and Issues

TODO

- [x] Add tests involving adding multiple keys at once
- [x] Rename `x/pkr` to `x/pubkey`
- [x] CLI test

Closes #162

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -245,7 +244,7 @@ type App struct {
 	// SEDA modules keepers
 	WasmStorageKeeper wasmstoragekeeper.Keeper
 	TallyKeeper       tallykeeper.Keeper
-	PkrKeeper         pkrkeeper.Keeper
+	PkrKeeper         pubkeykeeper.Keeper
```
Comment (Thomasvdam) [inline app/app.go:247]: Should this not also become `PubKeyKeeper`?

File: app/utils/vrf_key.go [hunk 2]
```diff
@@ -206,20 +206,15 @@ func LoadVRFKey(keyFilePath string) (*VRFKey, error) {
 // If loadPath is specified, it loads the VRF key file at the specified
 // path. Otherwise, it generates a new VRF key, whose entropy is randomly
 // generated or obtained from the mnemonic, if provided.
-func LoadOrGenVRFKey(config *cfg.Config, loadPath, mnemonic string) (vrfPubKey sdkcrypto.PubKey, err error) {
+func LoadOrGenVRFKey(config *cfg.Config, loadPath string) (vrfPubKey sdkcrypto.PubKey, err error) {
```
Comment (Thomasvdam) [inline app/utils/vrf_key.go:209]: I'm not 100% sure how the dependency between the PKR and other modules is going to look. I thought the pubkeys was going to generate all required keys and the modules that utilise them 'just' send data to the module to have it signed with the correct key.

File: x/pubkey/client/cli/tx.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		AddKey(valAddrCodec),
+	)
+	return cmd
+}
+
+// AddKey returns the command for adding a new key and uploading its
+// public key on chain at a given index.
+func AddKey(ac address.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "add-key",
```
Comment (Thomasvdam) [inline x/pubkey/client/cli/tx.go:50]: Maybe we should update the name and descriptions to indicate it generates all keys and submits a TX to register the new public keys?

File: x/pubkey/keeper/msg_server.go [hunk 4]
```diff
@@ -39,22 +39,24 @@ func (m msgServer) AddKey(goCtx context.Context, msg *types.MsgAddKey) (*types.M
...
-		return nil, sdkerrors.ErrInvalidType.Wrapf("%T is not a cryptotypes.PubKey", pubKey)
+	for _, indPubKey := range msg.IndexedPubKeys {
+		pubKey, ok := indPubKey.PubKey.GetCachedValue().(cryptotypes.PubKey)
+		if !ok {
+			return nil, sdkerrors.ErrInvalidType.Wrapf("%T is not a cryptotypes.PubKey", pubKey)
+		}
+
+		if err := m.Keeper.PubKeys.Set(ctx, collections.Join(valAddr, indPubKey.Index), pubKey); err != nil {
+			return nil, err
+		}
```
Comment (Thomasvdam) [inline x/pubkey/keeper/msg_server.go:50]: For a different PR: move the knowledge of the store/collections out of the message server?

File: x/pubkey/types/errors.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package types
+
+import "cosmossdk.io/errors"
+
+var (
+	ErrEmptyValue        = errors.Register("pubkey", 1, "empty value")
+	ErrInvalidType       = errors.Register("pubkey", 2, "invalid type")
+	ErrValidatorNotFound = errors.Register("pubkey", 3, "validator not found")
```
Comment (Thomasvdam) [inline x/pubkey/types/errors.go:8]: Do these custom errors add value over the SDK default errors or can we simply wrap a standard error with a description?

**Review by Thomasvdam — APPROVED** [review]

**Comment by hacheigriega** [comment]
Thanks for the reviews.. I will address them after fixing the JSON unmarshalling issue I found while adding CLI tests.

**Comment by hacheigriega** [comment]
Done but there is a gas estimation issue to be addressed in a separate PR (See Issue #341)

**Pull request merged**